sql: segregate query plan preparation in a new expandPlan interface.

knz opened the pull request:
##The makePlan() interface was not well designed, it had too much
responsibility. There are really 6 semantic phases on SQL statements:
1. structural syntax checking.
2. gathering the table descriptors and resolving qualified names.
3. type inference and type checking.
4. normalization and substituting placeholders with actual values.
5. building the query plan, including index selection.
6. running the query plan.

The pgwire Prepare phase needs to do steps 1-3 only.
The EXPLAIN statement needs to do steps 1-5 only.

Prior to this patch, the makePlan() interface does steps 1-5 in one
go. This is arguably too much, although the overhead is not visible to
users.

Prior to this patch, the Start() interface re-did steps 2 to 5 again,
and also does step 6 for some statements, in particular DELETE which
has a "fast path" for when there is no WHERE clause or index updates.

Now EXPLAIN needs a full query plan to work properly, ie it needs to
observe completion of step 5. This means EXPLAIN could not merely use
makePlan() and had to call Start() to get the fully build query
plan. But then because Start() also performs side effects (eg. for
DELETE), this would cause EXPLAIN to do too much work, in particular
delete data for DELETE. This is a bug.

This patch addresses this issue by clearly separating:
- the role of newPlan(), which is to achieve steps 1-3, in some cases step 4.
- a new role played by a new interface planNode.expandPlan(), which is
  to achieve step 5, possibly re-running step 4, without ever starting
  step 6.
- the role of Start(), which may start step 6.

Then makePlan() invokes newPlan() and then expandPlan().

Fixes #6613.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6622)

<!-- Reviewable:end -->

knz committed:
Make distinctNode a fully-fledged planNode.

Prior to this patch distinctNode would (ab)use the Go implicit
delegation syntax to automatically import the methods of its
sub-node. This makes reading the code slightly less straightforward
(nearly all other nodes use an explicit delegation, so this comes as
an exception) and makes debugging objectively harder since the
distinctNode disappears from the stack trace entirely for the methods
so "inherited".

This patch alleviates the issue by ensuring that distinctNode
implements all the planNode interface itself.

knz commented:
cc @nvanbenschoten @RaduBerinde.

RaduBerinde commented:
The change is missing proper documentation for `makePlan`. We need a description of what exactly it returns (something that AFAICT is completely undefined at this point, even in the commit message), and we should mention the expectation that `BuildPlan` is to be called on the result.

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> cc @nvanbenschoten @RaduBerinde.</details>


---

Review status: 0 of 22 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/plan.go, line 431 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHUxi5mP71oF0WNdLAI:-KHUxi5mP71oF0WNdLAJ:-1635371622) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc9658a28971d456a2dbd7495756c870c1a81512/sql/plan.go#L431)):*
> ```Go
> 	DebugValues() debugValues
> 
> 	// BuildPlan finalizes type checking of placeholders and builds
> ```

"build a plan" completely ignores the fact that it is being run on a plan node/tree. The doc should explain how the plan it is run on **changes**. 

---

*[sql/plan.go, line 434 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHUxu6STKrukHzniJH-:-KHUxu6STKrukHzniJH0:1660845035) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc9658a28971d456a2dbd7495756c870c1a81512/sql/plan.go#L434)):*
> ```Go
> 	// the query plan. Returns an error if the initialization fails.
> 	// The SQL "prepare" phase, as well as the EXPLAIN statement, should
> 	// merely build the plan node(s) and call BuildPlan().
> ```

Related to the above - "build the plan node(s)" here is not really defined. We are saying that BuildPlan builds the plan..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHUuzWsRT_fnWEJ0qFz:1285194270)*
<!-- Sent from Reviewable.io -->

knz committed:
Move the empty planNode to its own source file.

knz committed:
Make the limitNode a fully fledged planNode.

Prior to this patch limitNode would (ab)use the Go implicit delegation
syntax to automatically import the methods of its sub-node. This makes
reading the code slightly less straightforward (nearly all other nodes
use an explicit delegation, so this comes as an exception) and makes
debugging objectively harder since the limitNode disappears from the
stack trace entirely for the methods so "inherited".

This patch alleviates the issue by ensuring that limitNode
implements all the planNode interface itself.

knz committed:
Improve and better document the source layout of the planner.

NB: This patch only moves code around to different files, changes
comments and adds non-functional interfaces (= no logic changes).

The "planner" object in the sql package is a bit a showcase of the
"God class" antipattern. This patch alleviates this situation by
re-organizing the services of the planner in separate files that
highlight its separate roles:

- its role as manager of DB descriptors, via a DescriptorAccessor
  interface which documents all the related methods (in
  descriptor.go).

- its role as accessor for database descriptors via a DatabaseAccessor
  interface which documents all the related methods (in database.go).

- its role as accessor for table descriptors and table schema updates,
  via a SchemaAccessor interface which documents all the related
  methods (in table.go).

The planner constructor and struct definition is moved away from
`plan.go` into a new `planner.go`. `plan.go` remains to define the
planNode interface and assert its implementation by the various
concrete structs.

The methods corresponding to each sub-interface are also moved to the
primary source file for that interface.

knz commented:
@RaduBerinde agreed. I have improved the documentation in the comments accordingly. Also the interface is now named "FinalizePlan" for additional clarity.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The change is missing proper documentation for `makePlan`. We need a description of what exactly it returns (something that AFAICT is completely undefined at this point, even in the commit message), and we should mention the expectation that `BuildPlan` is to be called on the result.</details>


---

Review status: 0 of 27 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 431 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHUxi5mP71oF0WNdLAI:-KHVLuoFFik5oYcAPn5Q:-2027441553) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc9658a28971d456a2dbd7495756c870c1a81512/sql/plan.go#L431)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> "build a plan" completely ignores the fact that it is being run on a plan node/tree. The doc should explain how the plan it is run on **changes**. </details>
Agreed. See the new comments.

---

*[sql/plan.go, line 434 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHUxu6STKrukHzniJH-:-KHVLz9HvsfrN0PPqQ5M:-2027441553) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc9658a28971d456a2dbd7495756c870c1a81512/sql/plan.go#L434)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Related to the above - "build the plan node(s)" here is not really defined. We are saying that BuildPlan builds the plan..</details>
Agreed. See the new comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVLfmA-nh3nJwM4MrK:-800056460)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<meta> I suggest we always rebase on a fresh master when exporting PRs with many commits like this, otherwise it's hard to tell which commits are obsolete and which aren't. (if you rebase on a new master, all the current commits are gonna be a suffix)

I understand why we need to have a way to perform just 1-3 (for prepare), but why does `Finalize` need to be a method on a node? Why can't it be embedded in `makePlan` (but not in `planner.prepare()`)?

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> @RaduBerinde agreed. I have improved the documentation in the comments accordingly. Also the interface is now named "FinalizePlan" for additional clarity.</details>


---

Review status: 0 of 27 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/database.go, line 95 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVSKqKQL3ofW1Y56ZZ:-KHVSKqKQL3ofW1Y56Z_:-613620377) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/database.go#L95)):*
> ```Go
> 
> // DatabaseAccessor provides helper methods for using SQL database descriptors.
> type DatabaseAccessor interface {
> ```

I don't know if I like all these new interfaces? What's the point, if they're still implemented by planner? You still have to look in a bunch of files to see all the things that the does. I don't think they help with segmenting anything off, they could be replaced with comments. 
How about making these interfaces actual structs, sub-objects of the planner? Then you could say you've done something to split responsibilities better.

---

*[sql/plan.go, line 85 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVR-ZWCczd2K1wCah1:-KHVR-ZX1LFIebSYuvDP:177152323) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/plan.go#L85)):*
> ```Go
> }
> 
> type planNodeFastPath interface {
> ```

can you hint as to who's gonna implement this and how it's used? Can't it be part of `planNode`?

---

*[sql/plan.go, line 25 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVXGmcz4QUtDVGX5V8:-KHVXGmdk_IbPHaGvhvU:-995654820) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L25)):*
> ```Go
> )
> 
> type planMaker interface {
> ```

again, I don't know if this interface accomplishes anything. But it'd be nice if these two methods were isolated to some singleton (which should be called `planner` or `planMaker`, different from everything else on the current `planner` that's used during query execution.

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVVHQ7JjfFlETX51IV:-KHVVHQ7JjfFlETX51IW:-767966175) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L39)):*
> ```Go
> 	// Note: The autoCommit parameter enables operations to enable the
> 	// 1PC optimization. This is a bit hackish/preliminary at present.
> 	makePlan(stmt parser.Statement, desiredTypes []parser.Datum, autoCommit bool) (planNode, error)
> ```

can you suggest why `FinalizePlan()` is not part of `makePlan()`? Is it because "prepare" doesn't need it?
If so, it seems that `autoCommit` should be passed to `FinalizedPlan()` and not to this guy? Or even to `Start()`?
Or does it have something to do with subqueries? If so, is there a good reason why we can't handle subqueries during `makePlan` but we can during `Finalize`?

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVXqa8YlMJE-W-UetH:-KHVXqa8YlMJE-W-UetI:-1719667600) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L39)):*
> ```Go
> 	// Note: The autoCommit parameter enables operations to enable the
> 	// 1PC optimization. This is a bit hackish/preliminary at present.
> 	makePlan(stmt parser.Statement, desiredTypes []parser.Datum, autoCommit bool) (planNode, error)
> ```

can we please add a comment for `desiredTypes`

---

*[sql/plan.go, line 46 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVUzQBaGAGwSh_xXAy:-KHVUzQCdc7wWjW_rBGc:74462271) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L46)):*
> ```Go
> 	// will do just enough work to check the structural validity of the
> 	// SQL statement and determine types for placeholders. However it is
> 	// not appropriate to call BuildPlan(), Next() or Values() on a plan
> ```

s/BuildPlan/FinalizePlan ?

---

*[sql/planner.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVT4O9_d_FDe-jAVSO:-KHVT4O9_d_FDe-jAVSP:490174569) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/planner.go#L70)):*
> ```Go
> // Planner abstracts the services provided by a planner object
> // to the other SQL front-end components.
> type Planner interface {
> ```

what's up with this (new?) interface? Why do we need it / why does it need to be public?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVTf9ZZc0LMQ6Xk2iI:1793265465)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 1 files at r1, 22 of 22 files at r2, 2 of 2 files at r3, 23 of 23 files at r4, 1 of 1 files at r5, 5 of 5 files at r6, 3 of 26 files at r8, 24 of 24 files at r9.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[sql/database.go, line 91 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVZN9iatIpXW8WQ5MA:-KHVZN9jnwBvbQpVNFD-:1429205788) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/database.go#L91)):*
> ```Go
> 	nameKey = sqlbase.MakeNameMetadataKey(keys.RootNamespaceID, dbDesc.GetName())
> 	descKey = sqlbase.MakeDescMetadataKey(dbDesc.ID)
> 	return
> ```

"naked" returns are generally discouraged in Go unless there is a particularly good reason for them.

---

*[sql/limit.go, line 98 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVYUWqxr96JhMpfGzg:-KHVYUWrNgGr-r5bda3q:-1257895470) ([raw file](https://github.com/cockroachdb/cockroach/blob/15e90b83d2fbe370411686ebacbda0d01db8c31c/sql/limit.go#L98)):*
> ```Go
> 
> type limitNode struct {
> 	plan      planNode
> ```

Why un-embed the `planNode` here and in a few other places if you're still going to defer primarily to the wrapped `planNode`?

---

*[sql/plan.go, line 438 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVWiBcbqh6i8u6nZkZ:-KHVWiBdu-9ZrFfS5Ku5:835679039) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ca63916cd8a404d977d1693e003ba401979a839/sql/plan.go#L438)):*
> ```Go
> 
> 	// Start begins the processing of the query/statement and starts
> 	// performing side effects for data-modifying statements. Returns an
> ```

I like how this comment specifies that `Start` will have side effects. You might want to add to the contract above that the others will not have side effects.

---

*[sql/table.go, line 101 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVZjJWple_U-HIjuWW:-KHVZjJWple_U-HIjuWX:-1685359659) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/table.go#L101)):*
> ```Go
> 
> // SchemaAccessor provides helper methods for using the SQL schema.
> type SchemaAccessor interface {
> ```

What is the point of defining these interfaces if they are never used for indirection? Just to logically group behavior together?

---

*[sql/update.go, line 292 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVaDWqEoj2qSWLenUJ:-KHVaDWsyCidR-t_s8lc:276190340) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/update.go#L292)):*
> ```Go
> 
> 	// Create the workhorse select clause for rows that need updating.
> 	// TODO(knz): See comment above in Update().
> ```

Will this change enable us to clean this second `SelectClause` as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622)*
<!-- Sent from Reviewable.io -->

knz commented:
There are only 2 commits in this PR. I will try rebasing to clean up the mess as you suggest.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> <meta> I suggest we always rebase on a fresh master when exporting PRs with many commits like this, otherwise it's hard to tell which commits are obsolete and which aren't. (if you rebase on a new master, all the current commits are gonna be a suffix)
> 
> I understand why we need to have a way to perform just 1-3 (for prepare), but why does `Finalize` need to be a method on a node? Why can't it be embedded in `makePlan` (but not in `planner.prepare()`)?</details>


---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[sql/database.go, line 91 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVZN9iatIpXW8WQ5MA:-KHVczdoj5uqB58FZ60T:922457833) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/database.go#L91)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> "naked" returns are generally discouraged in Go unless there is a particularly good reason for them.</details>
I'll queue this for a separate commit. This commit is only rearranging comments/methods and does not change any code. Let's keep it like that.

---

*[sql/database.go, line 95 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVSKqKQL3ofW1Y56ZZ:-KHVd6zLPXw0fuogDB3a:-987595909) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/database.go#L95)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I don't know if I like all these new interfaces? What's the point, if they're still implemented by planner? You still have to look in a bunch of files to see all the things that the does. I don't think they help with segmenting anything off, they could be replaced with comments. 
> How about making these interfaces actual structs, sub-objects of the planner? Then you could say you've done something to split responsibilities better.</details>
The interface is introduced purely for documentation purposes. It helps by grouping all the method comments in one place so they can be groked more quickly in the code editor. It has nothing to do with code semantics.

---

*[sql/limit.go, line 98 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVYUWqxr96JhMpfGzg:-KHVdJBEVWoaoaM0SjSY:-652033329) ([raw file](https://github.com/cockroachdb/cockroach/blob/15e90b83d2fbe370411686ebacbda0d01db8c31c/sql/limit.go#L98)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Why un-embed the `planNode` here and in a few other places if you're still going to defer primarily to the wrapped `planNode`?</details>
Because it makes the code more regular and easier to read.

---

*[sql/plan.go, line 438 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVWiBcbqh6i8u6nZkZ:-KHVc6F6RRxXZb9h1qyW:-486965676) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ca63916cd8a404d977d1693e003ba401979a839/sql/plan.go#L438)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I like how this comment specifies that `Start` will have side effects. You might want to add to the contract above that the others will not have side effects.</details>
I think this is amply documented by the existing comments already. 

---

*[sql/plan.go, line 85 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVR-ZWCczd2K1wCah1:-KHVbn51tVNgeSosx80z:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/plan.go#L85)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you hint as to who's gonna implement this and how it's used? Can't it be part of `planNode`?</details>
Done.

---

*[sql/plan.go, line 25 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVXGmcz4QUtDVGX5V8:-KHVakb_3NILoAUEC2wS:-1653836710) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L25)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> again, I don't know if this interface accomplishes anything. But it'd be nice if these two methods were isolated to some singleton (which should be called `planner` or `planMaker`, different from everything else on the current `planner` that's used during query execution.</details>
This interface is introduced purely for documentation purposes. A later PR may push this agenda further doing what you suggest.

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVVHQ7JjfFlETX51IV:-KHVavbaSEFoKSRmUwUP:-1387530488) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L39)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you suggest why `FinalizePlan()` is not part of `makePlan()`? Is it because "prepare" doesn't need it?
> If so, it seems that `autoCommit` should be passed to `FinalizedPlan()` and not to this guy? Or even to `Start()`?
> Or does it have something to do with subqueries? If so, is there a good reason why we can't handle subqueries during `makePlan` but we can during `Finalize`?</details>
You are asking too many questions :)
Why FinalizePlan() is not part of makePlan() -> because prepare does not need it, indeed.

autoCommit can be moved, you are right, but this would belong to a different commit (which I could implement as a followup, I'll see).

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVXqa8YlMJE-W-UetH:-KHVbf-oBUALBPadmPpG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L39)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can we please add a comment for `desiredTypes`</details>
Done.

---

*[sql/plan.go, line 46 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVUzQBaGAGwSh_xXAy:-KHVbl9ps3FCge1aZnJc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/plan.go#L46)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/BuildPlan/FinalizePlan ?</details>
Done.

---

*[sql/planner.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVT4O9_d_FDe-jAVSO:-KHVcDOCWr-odB04nD_c:-1055012730) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/planner.go#L70)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what's up with this (new?) interface? Why do we need it / why does it need to be public?</details>
it does not need to be public. However I cannot use the name "planner" in lowercase because the struct already has it. Feel free to suggest a better name.

The interface is otherwise introduced purely for documentation purposes.

---

*[sql/table.go, line 101 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVZjJWple_U-HIjuWW:-KHVcPQuRxT2M5jLCPqz:761972571) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/table.go#L101)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> What is the point of defining these interfaces if they are never used for indirection? Just to logically group behavior together?</details>
The interface is introduced purely for documentation purposes. It groups all the method comments in one place, making them easy to read in the code editor.

---

*[sql/update.go, line 292 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVaDWqEoj2qSWLenUJ:-KHVcZHDILUoC2ajM1Of:-1186494905) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd9c35a29aa7d6e0bd4a74731e9687d194e8f62/sql/update.go#L292)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Will this change enable us to clean this second `SelectClause` as well?</details>
Yeah basically all this is a prerequisite for this other PR I have in the stove already (#6594)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVcj7l1nnG5npTtbqQ:195168416)*
<!-- Sent from Reviewable.io -->

knz committed:
Separate the makePlan() interface from the newPlan() constructor.

- makePlan() is used by the Executor and other places to translate a
  statement to a query plan.

- newPlan() is used recursively by the plan node constructors for
  intermediate syntax nodes.

RaduBerinde commented:
Review status: 0 of 27 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[sql/plan.go, line 60 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVlVS66RA7Uaxe8FqA:-KHVlVS66RA7Uaxe8FqB:-1424435559) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L60)):*
> ```Go
> 	//
> 	// Available after makePlan(). (FIXME: also prepare()? )
> 	Columns() []ResultColumn
> ```

index selection could affect `Columns()` if we are looking at an internal node. E.g. `SELECT y .. ORDER BY x`. If we use an index on `x`, the `selectNode` won't need to return `x`. I'm not sure if we are referring here to just the root of the tree, which indeed I don't think can change after `makePlan`.

---

*[sql/plan.go, line 64 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVkv4x2Atb_ZoLbWDl:-KHVkv4yM9mp3K-LAYAP:-1197707441) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L64)):*
> ```Go
> 	// The indexes of the columns the output is ordered by.
> 	//
> 	// Available after makePlan(). (FIXME: also prepare()? )
> ```

This may depend on the index, so it's not really available until after `FinalizePlan`. Or maybe it is available but with the caveat that it might change.

---

*[sql/plan.go, line 91 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVkUeWiDew9hND_eKw:-KHVkUeWiDew9hND_eKx:764792869) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L91)):*
> ```Go
> 	//
> 	// Available after makePlan().
> 	FinalizePlan() error
> ```

Thanks for the new comments! They look great. Also agree with the `FinalizePlan` name.

One suggestion: could we reorder the functions according to how "early" they are available? First `FinalizePlan` then `Columns' (?), 'Ordering', 'Explain*`, then 'MarkDebug', 'SetLimitHint`, then `Start` then `Next`,  then `Values/Debug`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622)*
<!-- Sent from Reviewable.io -->

knz commented:
I changed the interface name again, this time to expandPlan(), thanks to a suggestion by @petermattis that the word "Finalize" reminds the reader of finalizers and suggests the method is to be called after the node is not used any more.

Also I implemented @andreimatei's suggestions to make expandPlan() invisible to the users of makePlan().

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> There are only 2 commits in this PR. I will try rebasing to clean up the mess as you suggest.</details>


---

Review status: 0 of 27 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 60 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVlVS66RA7Uaxe8FqA:-KHVnCPaQdZAAKAkzGJs:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> index selection could affect `Columns()` if we are looking at an internal node. E.g. `SELECT y .. ORDER BY x`. If we use an index on `x`, the `selectNode` won't need to return `x`. I'm not sure if we are referring here to just the root of the tree, which indeed I don't think can change after `makePlan`.</details>
Done.

---

*[sql/plan.go, line 64 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVkv4x2Atb_ZoLbWDl:-KHVnJgT9gTTYEJ_Ji7k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L64)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> This may depend on the index, so it's not really available until after `FinalizePlan`. Or maybe it is available but with the caveat that it might change.</details>
Done.

---

*[sql/plan.go, line 91 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVkUeWiDew9hND_eKw:-KHVnvY5FDMz-pIMcJfS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd602251882e0f3dc934e334b1ecbdf1318c8a2/sql/plan.go#L91)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Thanks for the new comments! They look great. Also agree with the `FinalizePlan` name.
> 
> One suggestion: could we reorder the functions according to how "early" they are available? First `FinalizePlan` then `Columns' (?), 'Ordering', 'Explain*`, then 'MarkDebug', 'SetLimitHint`, then `Start` then `Next`,  then `Values/Debug`.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVo28GIvULTiDJrRAc:-1203350575)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Sshould we review only the last 2 of the 5 commits? Are the other three part of another PR (or they will be)?

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> There are only 2 commits in this PR. I will try rebasing to clean up the mess as you suggest.</details>


---

Review status: 0 of 27 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 58 \[r20\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVpArlOFhLrBpYmRC5:-KHVpArlOFhLrBpYmRC6:725532138) ([raw file](https://github.com/cockroachdb/cockroach/blob/41b364c74c764560d1232387138c481a5e508725/sql/plan.go#L58)):*
> ```Go
> 	// the result column types.
> 	//
> 	// Available after newPlan().
> ```

newPlan isn't defined in these interfaces.. Even in the code it's not really documented, it just says "newPlan constructs a planNode from a statement"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVmeHoc2_7qB-hQ6RX:-848109128)*
<!-- Sent from Reviewable.io -->

knz commented:
No actually all these commits belong here. You can review them separately though.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Sshould we review only the last 2 of the 5 commits? Are the other three part of another PR (or they will be)?</details>


---

Review status: 0 of 27 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 58 \[r20\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVpArlOFhLrBpYmRC5:-KHVr4SRaCXbB0ofbqM8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/41b364c74c764560d1232387138c481a5e508725/sql/plan.go#L58)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> newPlan isn't defined in these interfaces.. Even in the code it's not really documented, it just says "newPlan constructs a planNode from a statement"</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVpy_qdGjKiPGyNVfs:-2141439256)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ok cool.

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> No actually all these commits belong here. You can review them separately though.</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHVrtIkmyKEIz3K4wtH:-378852634)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Ok cool.</details>


---

Reviewed 1 of 27 files at r15, 3 of 5 files at r25, 23 of 23 files at r26, 1 of 1 files at r27.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 45 \[r26\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_AFPLkFgyKSCjqe8F:-KH_AFPLkFgyKSCjqe8G:-1057784589) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3820ee6eb4f0f72b377cd427d0a3922f2853e91/sql/plan.go#L45)):*
> ```Go
> 	// will do just enough work to check the structural validity of the
> 	// SQL statement and determine types for placeholders. However it is
> 	// not appropriate to call BuildPlan(), Next() or Values() on a plan
> ```

s/BuildPlan/expandPlan/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KH_B6sPIeUwdCbWJ-f8:1417088721)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[sql/plan.go, line 45 \[r26\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_AFPLkFgyKSCjqe8F:-KH_CN90ov2z03D-NJhW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3820ee6eb4f0f72b377cd427d0a3922f2853e91/sql/plan.go#L45)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> s/BuildPlan/expandPlan/</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622)*
<!-- Sent from Reviewable.io -->

tamird commented:
Please include rationale in the following commit messages:
- Make distinctNode a fully-fledged planNode.
- nMake the limit node a fully fledged planNode.

The interfaces-for-planner commit is incredibly hard to review. Is it all code movement? the commit message suggests there is some refactoring involved, but the size of the diff makes it quite hard to spot the relevant parts. I'm also :-1: on that commit generally for reasons pointed out by @andreimatei.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> :lgtm:</details>


---

Reviewed 1 of 27 files at r15, 26 of 26 files at r28, 1 of 1 files at r29, 2 of 2 files at r30, 1 of 1 files at r31, 5 of 5 files at r32, 23 of 23 files at r33.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/empty.go, line 39 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KAsUaimTr4R_SFnI:-KH_KAsUaimTr4R_SFnJ:1507100560) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/empty.go#L39)):*
> ```Go
> func (*emptyNode) MarkDebug(_ explainMode)             {}
> 
> // func (*emptyNode) BuildPlan() error                    { return nil }
> ```

nit: remove the addition of this code/comment from whichever commit added it

---

*[sql/limit.go, line 108 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KO30lWSXbai8q6eR:-KH_KO30lWSXbai8q6eS:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/limit.go#L108)):*
> ```Go
> func (n *limitNode) ExplainTypes(f func(string, string)) { n.plan.ExplainTypes(f) }
> 
> // func (n *limitNode) BuildPlan() error                    { return n.plan.BuildPlan() }
> ```

ditto

---

*[sql/planner.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVT4O9_d_FDe-jAVSO:-KH_JRBLhCkvR2Up0wtX:1559310967) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/planner.go#L70)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> it does not need to be public. However I cannot use the name "planner" in lowercase because the struct already has it. Feel free to suggest a better name.
> 
> The interface is otherwise introduced purely for documentation purposes.</details>
you could name it `plannerI`, but I still think these interfaces don't add value.

---

*[sql/planner.go, line 196 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KQXsHu_6pY7wsd-g:-KH_KQXsHu_6pY7wsd-h:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/planner.go#L196)):*
> ```Go
> 		return nil, err
> 	}
> 	//if err := plan.BuildPlan(); err != nil {
> ```

ditto

---

*[sql/planner.go, line 224 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KRdCh0ov-zfxh5Ws:-KH_KRdCh0ov-zfxh5Wt:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/planner.go#L224)):*
> ```Go
> 		return 0, err
> 	}
> 	//if err := plan.BuildPlan(); err != nil {
> ```

ditto

---

*[sql/testdata/delete, line 112 \[r33\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KjtZ9vJudGr2xJDU:-KH_KjtZ9vJudGr2xJDV:1567619500) ([raw file](https://github.com/cockroachdb/cockroach/blob/72304260ecdbada08c02a7ce61cac7c5dfb31dbd/sql/testdata/delete#L112)):*
> ```
> 
> query ITT colnames
> EXPLAIN DELETE FROM unindexed
> ```

what's being tested here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KH_I2bzCi4jeWeFCZPC:2114429884)*
<!-- Sent from Reviewable.io -->

knz committed:
Segregate query plan preparation in a new expandPlan interface.

The makePlan() interface was not well designed, it had too much
responsibility. There are really 6 semantic phases on SQL statements:

1.  structural syntax checking.
2.  gathering the table descriptors and resolving qualified names.
3.  type inference and type checking.
4.  normalization and substituting placeholders with actual values.
5.  building the query plan, including index selection.
6.  running the query plan.

The pgwire Prepare phase needs to do steps 1-3 only.
The EXPLAIN statement needs to do steps 1-5 only.

Prior to this patch, the makePlan() interface does steps 1-5 in one
go. This is arguably too much, although the overhead is not visible to
users.

Prior to this patch, the Start() interface re-did steps 2 to 5 again,
and also performed step 6 for some statements, in particular DELETE
which has a "fast path" for when there is no WHERE clause or index
updates.

Now EXPLAIN needs a full query plan to work properly, ie it needs to
observe completion of step 5. This means EXPLAIN could not merely use
makePlan() and had to call Start() to get the fully build query
plan. But then because Start() also performs side effects (eg. for
DELETE), this would cause EXPLAIN to do too much work, in particular
delete data for DELETE. This is a bug.

This patch addresses this issue by clearly separating:

- the role of newPlan(), which is to achieve steps 1-3, in some cases step 4.

- a new role played by a new interface planNode.expandPlan(), which is
  to achieve step 5, possibly re-running step 4, without ever
  starting step 6.

- the role of Start(), which may start step 6.

Using this, makePlan() invokes newPlan() then expandPlan(); EXPLAIN
uses makePlan() and prepare() uses only newPlan().

knz commented:
I have added the requested text in the commit messages. Also clarified the "interfaces-for-planner" commit that it does not contain any logic changes. I did not know that the word "refactoring" necessarily implies logic changes; I removed that word entirely from the commit summary.

I emphasize with the fact this is somewhat harder to review and I am willing to make steps to alleviate that. 

Meanwhile, introducing interfaces purely for documentation purposes has net benefits (improving code navigability and grokability), zero performance impact, so ...  I am strongly in favor of keeping this in and even extending the practice to other places. I gather that this looks & feels new to you but I still haven't heard arguments against.

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Please include rationale in the following commit messages:
> - Make distinctNode a fully-fledged planNode.
> - nMake the limit node a fully fledged planNode.
> 
> The interfaces-for-planner commit is incredibly hard to review. Is it all code movement? the commit message suggests there is some refactoring involved, but the size of the diff makes it quite hard to spot the relevant parts. I'm also :-1: on that commit generally for reasons pointed out by @andreimatei.
> 
> </details>


---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/empty.go, line 39 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KAsUaimTr4R_SFnI:-KH_U5K1vZ83YH4uC7ZO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/empty.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: remove the addition of this code/comment from whichever commit added it</details>
Done.

---

*[sql/limit.go, line 108 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KO30lWSXbai8q6eR:-KH_U6DJzP8IGm5rmjBh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/limit.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto</details>
Done.

---

*[sql/planner.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KHVT4O9_d_FDe-jAVSO:-KH_QPCpPaKPvtPvg8FL:-685666270) ([raw file](https://github.com/cockroachdb/cockroach/blob/04594b14c5fe91dbca4b928f9ba25c4412476203/sql/planner.go#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you could name it `plannerI`, but I still think these interfaces don't add value.</details>
ok I went for queryRunner which is closer to what this does anyways.

---

*[sql/planner.go, line 196 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KQXsHu_6pY7wsd-g:-KH_QSbuK3t7EX43W05E:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/planner.go#L196)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto</details>
Done.

---

*[sql/planner.go, line 224 \[r32\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KRdCh0ov-zfxh5Ws:-KH_QT8qpJmj0Xo5eQhO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0bc53b38146857b087f2c5ca1bacedfe41e1a0d/sql/planner.go#L224)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto</details>
Done.

---

*[sql/testdata/delete, line 112 \[r33\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-KH_KjtZ9vJudGr2xJDU:-KH_QnM2bFJb8H69zh6w:2099985452) ([raw file](https://github.com/cockroachdb/cockroach/blob/72304260ecdbada08c02a7ce61cac7c5dfb31dbd/sql/testdata/delete#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's being tested here?</details>
Added a comment to clarify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KH_T6olSAv5M9SnV9TX:-580917741)*
<!-- Sent from Reviewable.io -->

tamird commented:
Well, the primary argument against is the presence of more code which is effectively dead. Worse, the presence of this code may encourage people to use it, which _would_ have a performance impact.

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I have added the requested text in the commit messages. Also clarified the "interfaces-for-planner" commit that it does not contain any logic changes. I did not know that the word "refactoring" necessarily implies logic changes; I removed that word entirely from the commit summary.
> 
> I emphasize with the fact this is somewhat harder to review and I am willing to make steps to alleviate that. 
> 
> Meanwhile, introducing interfaces purely for documentation purposes has net benefits (improving code navigability and grokability), zero performance impact, so ...  I am strongly in favor of keeping this in and even extending the practice to other places. I gather that this looks & feels new to you but I still haven't heard arguments against.</details>


---

Reviewed 3 of 5 files at r37, 23 of 23 files at r38.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KH_VywQ17OmvqozBkGb:1214756135)*
<!-- Sent from Reviewable.io -->

knz commented:
The code is not dead because the next 2 lines assert that &planner{}
implement the interface. Should either the interface or the method
definitions change, the assertion would fail.

Then about using the interface instead of the object... I'm not sure
what the argument really is here. To me 1) it clearly does not make
sense to use this interface now 2) even if it did, the performance
impact of the interface indirection on the planner would be negligible
compared to that of, say, Statement or Expr 3) at some point in the
future we may want to split planner into multiple more independent
components, and it seems natural to me that whoever will make this step
will revisit these interfaces and refactor them towards their
refactoring accordingly.

What is the problem you are trying to prevent?

RaduBerinde commented:
I don't have an objection to the interfaces in this case. It's not a common pattern but I see it as a "blueprint" for how to later split planner into multiple components. I like that the functions are all documented in one place (and it won't bitrot like a huge comment or document).

I would be weary of extending this practice though, as it has a few disadvantages:
 - you have to write prototypes twice 
 - when you navigate code and jump to the implementation of a function, the documentation is not there (you have to go to the interface)
 - a first time reader will probably assume that there are multiple implementations to the interface.
 - it's reminiscent of C/C++ headers, not really consistent with go

Just my .02.

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> The code is not dead because the next 2 lines assert that &planner{} implement the interface. Should either the interface or the method definitions change, the assertion would fail.
> 
> Then about using the interface instead of the object... I'm not sure what the argument really is here. To me 1) it clearly does not make sense to use this interface now 2) even if it did, the performance impact of the interface indirection on the planner would be negligible compared to that of, say, Statement or Expr 3) at some point in the future we may want to split planner into multiple more independent components, and it seems natural to me that whoever will make this step will revisit these interfaces and refactor them towards their refactoring accordingly.
> 
> What is the problem you are trying to prevent? </details>


---

Review status: 4 of 27 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6622#-:-KHaK-6UxG18tmuL5srA:1796240148)*
<!-- Sent from Reviewable.io -->

knz commented:
@RaduBerinde thanks for the perspective. I think I agree; I will be careful not to push this further before we have more ample discussions on the topic. For now I'll merge this, and if we learn it was a terrible idea I'll volunteer to take the interfaces out later.
